File: python/sample_custom_commands.py
"""
Sample Custom Commands with Config Initialization
==================================================

LIFECYCLE:
    1. __init__(args) - Called first (args available here too, but prefer init)
    2. init(config)   - Called with full config including args - DO ALL SETUP HERE
    3. execute(client) - Called for each benchmark request

CONFIG PROPERTIES AVAILABLE IN init():
    config['custom_command_args']    - The --custom-command-args string (note: underscore not camelCase)
    config['data_size']              - Size of data in bytes (--datasize)
    config['keyspace_offset']        - Keyspace offset (--keyspace-offset)
    config['random_keyspace']        - Random keyspace size (--random)
    config['use_sequential']         - Whether sequential mode is enabled
    config['sequential_keyspacelen'] - Sequential keyspace length

EXAMPLE:
    python valkey-benchmark.py -t custom \
        --custom-command-file sample_custom_commands.py \
        --custom-command-args "operation=lpush_keyspace" \
        --sequential 1000000 --keyspace-offset 5000000
"""

import random
import string
from typing import Any, Dict, Optional


class CustomCommands:
    def __init__(self, args: Optional[str] = None):
        # Minimal constructor - all real setup happens in init()
        pass

    def init(self, config: Dict[str, Any]) -> None:
        """Initialize everything here - both custom args and benchmark config."""
        # Parse custom command args
        self.operation = 'set'
        self.batch_size = 1
        self.key_prefix = 'sample'
        self.counter = 0

        args = config.get('custom_command_args') or ''
        for pair in args.split(','):
            if '=' not in pair:
                continue
            key, value = pair.split('=', 1)
            key, value = key.strip(), value.strip()
            if key == 'operation':
                self.operation = value
            elif key == 'batch_size':
                self.batch_size = int(value) if value.isdigit() else 1
            elif key == 'key_prefix':
                self.key_prefix = value

        # Store benchmark config
        self.keyspace_offset = config.get('keyspace_offset', 0)
        self.random_keyspace = config.get('random_keyspace', 0)
        self.use_sequential = config.get('use_sequential', False)
        self.sequential_keyspacelen = config.get('sequential_keyspacelen', 0)
        self.data_size = config.get('data_size', 100)

        ks_type = 'sequential' if self.use_sequential else 'random' if self.random_keyspace > 0 else 'none'
        print(f'CustomCommands init: operation={self.operation}, keyspace={ks_type}, offset={self.keyspace_offset}')

    async def execute(self, client: Any) -> bool:
        ops = {
            'set': self._execute_set,
            'mset': self._execute_mset,
            'hset': self._execute_hset,
            'lpush': self._execute_lpush,
            'lpush_keyspace': self._execute_lpush_keyspace,
        }
        handler = ops.get(self.operation)
        return await handler(client) if handler else False

    def _get_next_key(self) -> str:
        if self.use_sequential:
            return f'key:{self.keyspace_offset + (self.counter % self.sequential_keyspacelen)}'
        elif self.random_keyspace > 0:
            return f'key:{self.keyspace_offset + random.randint(0, self.random_keyspace - 1)}'
        return f'key:{self.counter}'

    def _generate_data(self) -> str:
        return ''.join(random.choices(string.ascii_letters + string.digits, k=self.data_size))

    async def _execute_set(self, client: Any) -> bool:
        await client.set(f'{self.key_prefix}:key:{self.counter}', f'value:{self.counter}')
        self.counter += 1
        return True

    async def _execute_mset(self, client: Any) -> bool:
        kv_pairs = {}
        for _ in range(self.batch_size):
            kv_pairs[f'{self.key_prefix}:key:{self.counter}'] = f'value:{self.counter}'
            self.counter += 1
        await client.mset(kv_pairs)
        return True

    async def _execute_hset(self, client: Any) -> bool:
        await client.hset(f'{self.key_prefix}:hash', {f'field:{self.counter}': f'value:{self.counter}'})
        self.counter += 1
        return True

    async def _execute_lpush(self, client: Any) -> bool:
        await client.lpush(f'{self.key_prefix}:list', [f'value:{self.counter}'])
        self.counter += 1
        return True

    async def _execute_lpush_keyspace(self, client: Any) -> bool:
        key = self._get_next_key()
        data = self._generate_data()
        self.counter += 1
        await client.lpush(key, [data])
        return True

File: node/sample_custom_commands.js
/**
 * Sample Custom Commands with Config Initialization
 * ==================================================
 *
 * LIFECYCLE:
 *   1. constructor(args) - Called first (args available here too, but prefer init)
 *   2. init(config)      - Called with full config including args - DO ALL SETUP HERE
 *   3. execute(client)   - Called for each benchmark request
 *
 * CONFIG PROPERTIES AVAILABLE IN init():
 *   config.customCommandArgs       - The --custom-command-args string
 *   config.dataSize                - Size of data in bytes (--datasize)
 *   config.keyspaceOffset          - Keyspace offset (--keyspace-offset)
 *   config.randomKeyspace          - Random keyspace size (--random)
 *   config.useSequential           - Whether sequential mode is enabled
 *   config.sequentialKeyspacelen   - Sequential keyspace length
 *
 * EXAMPLE:
 *   node valkey-benchmark.js -t custom \
 *       --custom-command-file sample_custom_commands.js \
 *       --custom-command-args "operation=lpush_keyspace" \
 *       --sequential 1000000 --keyspace-offset 5000000
 */

class CustomCommands {
    constructor(args = null) {
        // Minimal constructor - all real setup happens in init()
    }

    /**
     * Initialize everything here - both custom args and benchmark config.
     * @param {Object} config - Full benchmark configuration
     */
    init(config) {
        // Parse custom command args
        this.operation = 'set';
        this.batchSize = 1;
        this.keyPrefix = 'sample';
        this.counter = 0;

        const args = config.customCommandArgs || '';
        for (const pair of args.split(',')) {
            const [key, ...rest] = pair.split('=');
            const value = rest.join('=').trim();
            if (key?.trim() === 'operation') this.operation = value;
            else if (key?.trim() === 'batch_size') this.batchSize = parseInt(value, 10) || 1;
            else if (key?.trim() === 'key_prefix') this.keyPrefix = value;
        }

        // Store benchmark config
        this.keyspaceOffset = config.keyspaceOffset || 0;
        this.randomKeyspace = config.randomKeyspace || 0;
        this.useSequential = config.useSequential || false;
        this.sequentialKeyspacelen = config.sequentialKeyspacelen || 0;
        this.dataSize = config.dataSize || 100;

        const ksType = this.useSequential ? 'sequential' : this.randomKeyspace > 0 ? 'random' : 'none';
        console.log(`CustomCommands init: operation=${this.operation}, keyspace=${ksType}, offset=${this.keyspaceOffset}`);
    }

    async execute(client) {
        switch (this.operation) {
            case 'set': return this._executeSet(client);
            case 'mset': return this._executeMset(client);
            case 'hset': return this._executeHset(client);
            case 'lpush': return this._executeLpush(client);
            case 'lpush_keyspace': return this._executeLpushKeyspace(client);
            default: return false;
        }
    }

    _getNextKey() {
        if (this.useSequential) {
            return `key:${this.keyspaceOffset + (this.counter % this.sequentialKeyspacelen)}`;
        } else if (this.randomKeyspace > 0) {
            return `key:${this.keyspaceOffset + Math.floor(Math.random() * this.randomKeyspace)}`;
        }
        return `key:${this.counter}`;
    }

    _generateData() {
        const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        let result = '';
        for (let i = 0; i < this.dataSize; i++) {
            result += chars.charAt(Math.floor(Math.random() * chars.length));
        }
        return result;
    }

    async _executeSet(client) {
        await client.set(`${this.keyPrefix}:key:${this.counter}`, `value:${this.counter++}`);
        return true;
    }

    async _executeMset(client) {
        const kvPairs = {};
        for (let i = 0; i < this.batchSize; i++) {
            kvPairs[`${this.keyPrefix}:key:${this.counter++}`] = `value:${this.counter}`;
        }
        await client.mset(kvPairs);
        return true;
    }

    async _executeHset(client) {
        await client.hset(`${this.keyPrefix}:hash`, { [`field:${this.counter++}`]: `value:${this.counter}` });
        return true;
    }

    async _executeLpush(client) {
        await client.lpush(`${this.keyPrefix}:list`, [`value:${this.counter++}`]);
        return true;
    }

    async _executeLpushKeyspace(client) {
        const key = this._getNextKey();
        const data = this._generateData();
        this.counter++;
        await client.lpush(key, [data]);
        return true;
    }
}

module.exports = CustomCommands;
